ASoC: SOF: ipc4: Support for ALSA switch/enum change notification from firmware

ujfalusi opened the pull request:
Hi,

This PR will add support to handle change notifications from firmware on ALSA switch/enum controls.
Similarly as IPC3 supported, firmware modules can update an ALSA control value depending on processing, like direction of arrival multi mic setups.

The defined way for the newly introduced kcontrol update is:
The notification is a MODULE_NOTIFICATION
The event_id is: upper 16 bit must be the magic value for ALSA control change notification (**0xA15A**), the lower 16 bit is the param_id of the control (same  value as used with large_config: 0x200 - switch, 0x201 - enum)
The event data must contain the `struct  sof_ipc4_control_msg_payload` either with the canv array or with no chanv data, but the id must be set correctly.

The kernel will update the control's values directly if chanv is sent alongside the notification or it will mark the control as dirty and the next read will be forced to fetch the updated values from the firmware.

plbossart commented on include/sound/sof/ipc4/header.h:
```diff
@@ -532,6 +532,35 @@ struct sof_ipc4_notify_resource_data {
 
+/**
+ * struct sof_ipc4_notify_module_data - payload for module notification
+ * @instance_id: isntance ID of the originator module of the notification
```
typo: instance

ujfalusi replied:
ok

plbossart replied:
not fixed @ujfalusi

ujfalusi replied:
aargh, sorry.

plbossart commented on include/sound/sof/ipc4/header.h:
```diff
@@ -532,6 +532,35 @@ struct sof_ipc4_notify_resource_data {
+	uint16_t module_id;
+	uint32_t event_id;
+	uint32_t event_data_size;
+	uint8_t event_data[];
```
is there a restriction on the event_data_size? Can we trust the value blindly or should there be some sanity-check?

ujfalusi replied:
There is no restriction, each module defines what is in here.
In this PR we are defining a generic switch/enum change notification.

I don't think there is a way to sanity-check this really, modules can do additional checks on the event_data.

plbossart commented on sound/soc/sof/ipc4-control.c:
```diff
@@ -240,6 +240,50 @@ sof_ipc4_set_generic_control_data(struct snd_sof_dev *sdev,
+	} else {
+		dev_err(scomp->dev, "Failed to read control data for %s\n",
+			scontrol->name);
+		scontrol->comp_data_dirty = true;
```
why is this an error? In the previous patch the notification data was tagged as optional, and the commit message also does not hint at any errors "If only a notification without values received, marked the control as dirty
and on next read access fetch the new values from the firmware."

ujfalusi replied:
This is not the notification handling function, this is a function to read the control data from the firmware with the GET_LARGE_CONFIG message.
Previously we did not needed this as the kernel knew the control's state as it was managing it, but with the notification support we need to be able to read the control's data if the notification did not contained the channel values.

plbossart replied:
not following at all, sorry.

ujfalusi replied:
When we receive a control change notification it can contain the new values for the control, in that case we will update the kcontrol based on the data received in the notification message.
The the new values for the control is optional, the firmware can send a notification with only the minimal information to identify the control, but no values are included.
In this case we mark the kcontrol as dirty and next time when the user reads the control we will use  GET_LARGE_CONFIG message to read the new values from firmware.
This error print is in this refresh code, it is not the notification handling path.

ranj063 commented on sound/soc/sof/ipc4-control.c:
```diff
@@ -601,6 +649,136 @@ sof_ipc4_volsw_setup(struct snd_sof_dev *sdev, struct snd_sof_widget *swidget,
+		break;
+	default:
+		dev_err(sdev->dev,
+			"%s: Invalid control type for module %u.%u: %u\n",
```
invalid param id?

ujfalusi replied:
OK, my reasoning was that the param_id for ALSA kcontrol support is the type of the control, it is a bit user friendly to print that.

ranj063 commented on sound/soc/sof/ipc4-control.c:
```diff
@@ -601,6 +649,136 @@ sof_ipc4_volsw_setup(struct snd_sof_dev *sdev, struct snd_sof_widget *swidget,
+	/* Find the swidget based on ndata->module_id and ndata->instance_id */
+	swidget = sof_ipc4_find_swidget_by_ids(sdev, ndata->module_id,
+					       ndata->instance_id);
+	if (!swidget) {
```
@ujfalusi I think we're missing a step during sof_widget_free() for IPC4 for this to work reliably. We need to invalidate the swidget->instance_id in sof_ipc4_widget_free()

ujfalusi replied:
Oh, I'll take a look, but the free path is only happening on probe error path and module removal, no?

ranj063 replied:
no free happens whenever the PCM is closed

ranj063 commented on sound/soc/sof/ipc4-control.c:
```diff
@@ -601,6 +649,136 @@ sof_ipc4_volsw_setup(struct snd_sof_dev *sdev, struct snd_sof_widget *swidget,
+				 * Mark the scontrol as dirty to force a refresh
+				 * on next read
+				 */
+				scontrol->comp_data_dirty = true;
```
can we do this outside the if {} on line 730 unconditionally?

ujfalusi replied:
No, it should not be unconditional. If the notification contains valid data then there is no need to reach out to the firmware, we got the changed value.

ujfalusi commented:
Changes since v1:
- typos fixed
- skip unused swidgets in `sof_ipc4_find_swidget_by_ids()` since only in use swidgets have valid instance_id
 - done this instead of a more intrusive instance_id management since it only matters in this case.

ranj063 reviewed (commented):
i'm good except for that one typo

ujfalusi commented:
Changes since v2:
- typo fixed in patch 2
